### ISSUE TITLE
[HOLD for payment 2023-03-01] [$1000] Allow avatar image uploads: `gif` on Android

### ISSUE THREAD

**Issue by Beamanator** [Expensify/App#14751]
# Action Performed:

1. Upload `.gif` as avatar image on Android

# Expected Result:

It should upload and convert to `.jpeg` (the conversion can happen on the front end if needed, but we already have conversion code on the backend that works for every type of image in web)

# Actual Result:

1. Uploading GIF in Android throws console warning, isn't able to upload image

Note: Some investigation was done [here](https://github.com/Expensify/App/pull/12549#issuecomment-1326579036), indicating this may be a problem with a library we use, `RNImageManipulator`

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~012d6d474213ddb534</li>
        <li>Upwork Job ID: 1621106096863707136</li>
        <li>Last Price Increase: 2023-02-10</li>
    </ul>
</details>

**Comment by s77rt** [comment]
### Proposal
I think the two mentioned issues have different root causes. In this proposal I'm handling the iOS `bmp` issue.
The `.bmp` images failing on iOS is a bug in `react-native-image-picker`. I have raised a [PR](https://github.com/react-native-image-picker/react-native-image-picker/pull/2076) that should fix the issue.

### RCA
Please refer to the PR:
https://github.com/react-native-image-picker/react-native-image-picker/pull/2076

**Comment by s77rt** [comment]
I have updated my proposal above to handle both issues.

**Comment by redstar504** [comment]
**Proposal (Android GIF issue)**

Modifying the upstream package with an additional dependency that decodes the entire GIF just to extract the first frame is not necessary. The existing decoder has an option built in that will capture the first frame as soon as it's available.

It is easy to implement, and saves memory that would be allocated to decoding the whole animation.  It can be accomplished by a simple change (in our existing [react-native-image-manipulator](https://github.com/Expensify/react-native-image-manipulator) fork):

```Java
diff --git a/android/src/main/java/com/reactnativeimagemanipulator/RNImageManipulatorModule.java b/android/src/main/java/com/reactnativeimagemanipulator/RNImageManipulatorModule.java
index 4b8ed8a..b404fa7 100644
--- a/android/src/main/java/com/reactnativeimagemanipulator/RNImageManipulatorModule.java
+++ b/android/src/main/java/com/reactnativeimagemanipulator/RNImageManipulatorModule.java
@@ -23,6 +23,7 @@ import com.facebook.react.bridge.ReactMethod;
 import com.facebook.react.bridge.ReadableArray;
 import com.facebook.react.bridge.ReadableMap;
 import com.facebook.react.bridge.WritableMap;
+import com.facebook.imagepipeline.common.ImageDecodeOptions;
 
 import java.io.ByteArrayOutputStream;
 import java.io.File;
@@ -53,6 +54,10 @@ public class RNImageManipulatorModule extends ReactContextBaseJavaModule {
         ImageRequestBuilder
             .newBuilderWithSource(Uri.parse(uriString))
             .setRotationOptions(RotationOptions.autoRotate())
+            .setImageDecodeOptions(
+                    ImageDecodeOptions.newBuilder()
+                            .setForceStaticImage(true)
+                            .build())
             .build();
     final DataSource<CloseableReference<CloseableImage>> dataSource
         = Fresco.getImagePipeline().fetchDecodedImage(imageRequest, getReactApplicationContext());

**Comment by s77rt** [comment]
@redstar504 Your solution is way better and straightforward, Good work!

@Beamanator I think we can ignore the upstream fix here https://github.com/oguzhnatly/react-native-image-manipulator/pull/3 since we no longer use that repo and we use our forked version (`package.json` was a bit confusing and I didn't catch that)

**Comment by s77rt** [comment]
@Beamanator Can you help us getting this https://github.com/react-native-image-picker/react-native-image-picker/pull/2076 merged asap? The maintainers are friendly and usually responsive.

**Comment by Beamanator** [comment]
First of all I'd like to see if @mananjadhav has the time to review y'all's proposals above 😅  Then we can work out which upstream PRs we are moving forward with 🙏

**Comment by trjExpensify** [comment]
@anmurali, I'm going to take this on from you as it's spawned from another issue I'm on related to image improvements.

**Comment by mananjadhav** [comment]
Thanks for the proposals folks. Sorry I was OOO and just got back to looking at the issues.

I think @redstar504's proposal for Android does work and good to go. question:

1. Are you submitting the proposal only for Android fix? What about iOS


@s77rt I checked your proposals too, question:

2. The change that you're recommending for iOS existed earlier and it was removed in this [PR](https://github.com/react-native-image-picker/react-native-image-picker/pull/1853). I don't see the reasoning mentioned (except the scope of PR is related Mac M1) and as you rightly mentioned that it is used for video. My findings are also that the same function is used to determine the type supporting even iOS 8+. So I think we should get some feedback there.

----

@Beamanator questions:

1. Because these two could be unrelated code changes, does it make sense to track them as separate issues? Concern is not only that it's unrelated but the proposals that look promising for both the platforms are by two contributors.

2. Should we ask to update the proposals with the new Proposal template?
3.

**Comment by redstar504** [comment]
@mananjadhav The proposed solution for iOS looks fine.  My proposal is only for the Android issue.  It may make sense to split this into two issues since they both have different root causes.  The iOS fix is so small that we might be able to apply it as a patch alongside the other small fixes in our /patches directory.

The only comment I can make on the proposed iOS change is that `kUTTypeImage` was deprecated in iOS 15.  I don't think that matters since they are using other deprecated types in the same file, and our build target is `>= iOS 11`.

The suggested condition `=> iOS 14` is:
``` ObjC
[self.result.itemProvider hasItemConformingToTypeIdentifier:UTTypeImage.identifier]
```

**Comment by Beamanator** [comment]
Created new issue to track the `iOS` fix: https://github.com/Expensify/App/issues/15035

> Should we ask to update the proposals with the new Proposal template?

That would be great - if there's old proposals not in the new format that's fine, but any new proposal should be in the new format 👍 🙏 

Also side note: @trjExpensify it looks like Melvin-bot auto-doubled this issue even though there's a lot of activity 😅 What should we do with that?

**Comment by trjExpensify** [comment]
Yep, proposals are under review correct?

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $1000

**Comment by mananjadhav** [comment]
> Yep, proposals are under review correct?

@trjExpensify Yes.


@Beamanator @redstar504's proposal for Android [here](https://github.com/Expensify/App/issues/14751#issuecomment-1420266690) is good.

**Comment by Beamanator** [comment]
@mananjadhav that proposal does look great, let's go for it 👍

**Comment by MelvinBot** [comment]
📣 @redstar504 You have been assigned to this job by @Beamanator!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by redstar504** [comment]
@Beamanator Thank you.  I've sent my Upwork application, and have opened a PR on the Expensify/react-native-image-manipulator fork here:  https://github.com/Expensify/react-native-image-manipulator/pull/2.

**Comment by mananjadhav** [comment]
@redstar504 Can you add the App PR with the updated commit hash?

**Comment by trjExpensify** [comment]
Sent offers to you both in the meantime as we review that PR. 👍

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.75-0 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15121


If no regressions arise, payment will be issued on 2023-03-01. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
📣 @MelvinBot! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@mananjadhav / @Beamanator] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@mananjadhav / @Beamanator] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@mananjadhav / @Beamanator] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@redstar504] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@trjExpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by mananjadhav** [comment]
@Beamanator @trjExpensify @redstar504 I think we're good to use the QA steps in the PR for the regression test proposal?

**Comment by redstar504** [comment]
@mananjadhav Yes, I will paste them here for reference:

https://github.com/Expensify/App/pull/15121

1. Add an animated GIF image to an Android simulator or device
2. Log in to the staging environment, and navigate to the profile screen
3. Press on the current avatar to change it, and press "Upload photo"
4. Press "Choose from gallery" and select the GIF image on your device
5. Proceed to crop, rotate, and zoom the avatar as desired
6. Press "Save" to finish uploading the avatar
7. Observe that the avatar was uploaded successfully
8. Observe that the saved avatar is a static image that only contains of the first frame of the animation

**Comment by mananjadhav** [comment]
Thanks @redstar504. Quick bump @Beamanator @trjExpensify

**Comment by Beamanator** [comment]
Those test steps look great! I'd say one thing to change is: don't mention the particular environment that the tests will be run in please 🙏

**Comment by redstar504** [comment]
@Beamanator Good catch. Edited.

**Comment by trjExpensify** [comment]
Do we want to test this specific file format on Android-only with every regression test? I.e why not across all platforms? Also, we don't have any file type specifics for avatar uploads in our regression suite atm, should we couple this with the most common? I.e JPEG, PNG, GIF, SVG? 

@Beamanator reference being somewhat [similar to this "other" regression](https://expensify.testrail.io/index.php?/cases/view/1977396&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=229062) test case for attachment uploads.

**Comment by Beamanator** [comment]
@trjExpensify Ooh good points! I think it's fair to make sure all of those types work across all platforms to prevent any bugs leaking through 👍

**Comment by trjExpensify** [comment]
Okay, so I think we add GIF, PNG and SVG to [this test case](https://expensify.testrail.io/index.php?/cases/view/1971220&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=229064) under a new sub-header:

**GIF, SVG and PNG**

1. Repeat the steps to upload an avatar, but now choose a PNG file within the file size limits
2. **Verify** that the avatar uploads successfully
3. Repeat the steps to upload an avatar, but now choose an SVG file 
4. **Verify** that the avatar uploads successfully
5. Repeat the steps to upload an avatar, but now choose a GIF file 
6. **Verify** that the avatar uploads successfully
7. **Verify** that the saved avatar is a static image that only contains the first frame of the animation

Let me know if you agree, and I'll create the GH.

**Comment by mananjadhav** [comment]
The test looks good.

>  within the file size limits

I don't have access to the Testrail but do we have this limit specified somewhere?

-----

@trjExpensify @Beamanator This is ready for payout today but I have a questions. While the review was done within the timelines (it had 2 PRs), it was merged a week later. Will the timeline bonus be affected with this?

**Comment by trjExpensify** [comment]
>I don't have access to the Testrail but do we have this limit specified somewhere?

Ah yeah, the earlier part of this test script is about file size/limits, so here I'm not rehashing that: 

<img width="566" alt="image" src="https://user-images.githubusercontent.com/16232057/222147119-d15348c2-5107-49dc-8023-b8ce94ca8987.png">

**Comment by mananjadhav** [comment]
thanks for posting the testcase.

**Comment by trjExpensify** [comment]
>@trjExpensify @Beamanator This is ready for payout today but I have a questions. While the review was done within the timelines (it had 2 PRs), it was merged a week later. Will the timeline bonus be affected with this?

Hey @mananjadhav, the bonus won't apply on this issue because the review wasn't completed until 4 days later. I see you approved the PR on the 16th, but the checklist wasn't completed until the 17th. Those two actions should come together for a review to be deemed complete. 

I've gone ahead and settled up the contracts. I'll create the regression test issue for Applause now and close this issue out. Thanks everyone!

**Comment by trjExpensify** [comment]
Regression test here: https://github.com/Expensify/Expensify/issues/265438

**Comment by mananjadhav** [comment]
Thanks for confirmation, but if you don’t mind, I would like to clarify (not for this issue but in general). Do we apply the same timeline of 3 business days when we have 2 PRs ?

**Comment by kbecciv** [comment]
@trjExpensify Can you please create the regression test issue or I can create one, please let me know.
QA team is unable to update GIF file in avatar on Android, Save button is not responding.

https://user-images.githubusercontent.com/93399543/230657686-5fb8472b-91b3-43f9-9fd6-90686a5b7bec.mp4

**Comment by trjExpensify** [comment]
Hey @kbecciv, sorry for the slight delay, I've been OoO. The regression test issue was created a while back [here](https://github.com/Expensify/Expensify/issues/265438), is that what you're looking for?

**Comment by trjExpensify** [comment]
Yep, please create a bug report for that 👍

**Comment by kbecciv** [comment]
@trjExpensify Logged separate issue https://github.com/Expensify/App/issues/18088

**Comment by trjExpensify** [comment]
Thanks!
